### PR TITLE
Stop mutating nodes in ColorConnected filter

### LINKED ISSUE #1192: Concurrent writes on map
Offending goroutine:

```
fatal error: concurrent map writes
goroutine 37 [running]:
runtime.throw(0x182bb50, 0x15)
        /usr/local/go/src/runtime/panic.go:530 +0x90 fp=0xc82d196f78 sp=0xc82d196f60
runtime.mapassign1(0x15de180, 0xc82794c780, 0xc82d197408, 0xc82d197880)
        /usr/local/go/src/runtime/hashmap.go:540 +0x54b fp=0xc82d197020 sp=0xc82d196f78
github.com/weaveworks/scope/render.ColorConnected.func1(0xc82794c780, 0xc825f78000)
        /go/src/github.com/weaveworks/scope/render/filters.go:50 +0x35f fp=0xc82d197b48 sp=0xc82d197020
github.com/weaveworks/scope/render.CustomRenderer.Render(0x1a14ad0, 0x7f708e2ff208, 0xc820399160, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, ...)
        /go/src/github.com/weaveworks/scope/render/filters.go:21 +0x70 fp=0xc82d197c20 sp=0xc82d197b48
github.com/weaveworks/scope/render.(*CustomRenderer).Render(0xc820388080, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, 0xc823f70480, 0xc823f704b0, ...)
        <autogenerated>:63 +0xc4 fp=0xc82d197cf8 sp=0xc82d197c20
github.com/weaveworks/scope/render.(*Filter).render(0xc82d198ab0, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, 0xc823f70480, 0xc823f704b0, ...)
        /go/src/github.com/weaveworks/scope/render/filters.go:78 +0xee fp=0xc82d1989c8 sp=0xc82d197cf8
github.com/weaveworks/scope/render.Filter.Stats(0x7f708e2ff268, 0xc820388080, 0x1a14b00, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, ...)
        /go/src/github.com/weaveworks/scope/render/filters.go:117 +0x4c fp=0xc82d198ab0 sp=0xc82d1989c8
github.com/weaveworks/scope/render.(*Filter).Stats(0xc8203880a0, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, 0xc823f70480, 0xc823f704b0, ...)
        <autogenerated>:65 +0xce fp=0xc82d198b88 sp=0xc82d198ab0
github.com/weaveworks/scope/render.(*memoise).Stats(0xc8203880c0, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, 0xc823f70480, 0xc823f704b0, ...)
        <autogenerated>:66 +0x83 fp=0xc82d198c60 sp=0xc82d198b88
github.com/weaveworks/scope/app.decorateWithStats(0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, 0xc823f70480, 0xc823f704b0, 0xc825855170, ...)
        /go/src/github.com/weaveworks/scope/app/api_topologies.go:235 +0x1c9 fp=0xc82d199060 sp=0xc82d198c60
github.com/weaveworks/scope/app.(*registry).renderTopologies.func1(0x1763450, 0x9, 0x0, 0x0, 0x7f708e2ff208, 0xc820388060, 0x174f660, 0x9, 0x1, 0xc8202bfce0, ...)
        /go/src/github.com/weaveworks/scope/app/api_topologies.go:215 +0x1c6 fp=0xc82d1991c0 sp=0xc82d199060
github.com/weaveworks/scope/app.(*registry).walk(0x23a5320, 0xc82d1995a8)
        /go/src/github.com/weaveworks/scope/app/api_topologies.go:191 +0x2bc fp=0xc82d199580 sp=0xc82d1991c0
github.com/weaveworks/scope/app.(*registry).renderTopologies(0x23a5320, 0xc825f78000, 0xc82182cc30, 0xc82182cc60, 0xc830e27680, 0xc830e276b0, 0xc8255a6330, 0xc82d23c480, 0xc823f70480, 0xc823f704b0, ...)
        /go/src/github.com/weaveworks/scope/app/api_topologies.go:218 +0xa7 fp=0xc82d1995d0 sp=0xc82d199580
github.com/weaveworks/scope/app.(*registry).makeTopologyList.func1(0x7f7090b4a068, 0xc826cccb70, 0x7f7090b49ff0, 0xc827d3f1c0, 0xc820107b20)
        /go/src/github.com/weaveworks/scope/app/api_topologies.go:203 +0x16d fp=0xc82d1997c8 sp=0xc82d1995d0
github.com/weaveworks/scope/app.requestContextDecorator.func1(0x7f7090b49ff0, 0xc827d3f1c0, 0xc820107b20)
        /go/src/github.com/weaveworks/scope/app/router.go:37 +0x226 fp=0xc82d1998b0 sp=0xc82d1997c8
net/http.HandlerFunc.ServeHTTP(0xc820339780, 0x7f7090b49ff0, 0xc827d3f1c0, 0xc820107b20)
        /usr/local/go/src/net/http/server.go:1618 +0x3a fp=0xc82d1998d0 sp=0xc82d1998b0
github.com/weaveworks/scope/vendor/github.com/PuerkitoBio/ghost/handlers.GZIPHandler.func1(0x7f7090b49ed8, 0xc828c93c70, 0xc820107b20)
        /go/src/github.com/weaveworks/scope/vendor/github.com/PuerkitoBio/ghost/handlers/gzip.go:127 +0x330 fp=0xc82d199968 sp=0xc82d1998d0
net/http.HandlerFunc.ServeHTTP(0xc820389420, 0x7f7090b49ed8, 0xc828c93c70, 0xc820107b20)
        /usr/local/go/src/net/http/server.go:1618 +0x3a fp=0xc82d199988 sp=0xc82d199968
github.com/weaveworks/scope/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc82037c190, 0x7f7090b49ed8, 0xc828c93c70, 0xc820107b20)
        /go/src/github.com/weaveworks/scope/vendor/github.com/gorilla/mux/mux.go:100 +0x29e fp=0xc82d199aa0 sp=0xc82d199988
github.com/weaveworks/scope/app.TopologyHandler.func1(0x7f7090b49ed8, 0xc828c93c70, 0xc820107b20)
        /go/src/github.com/weaveworks/scope/app/router.go:113 +0xa7 fp=0xc82d199b30 sp=0xc82d199aa0
net/http.HandlerFunc.ServeHTTP(0xc8203b07a0, 0x7f7090b49ed8, 0xc828c93c70, 0xc820107b20)
        /usr/local/go/src/net/http/server.go:1618 +0x3a fp=0xc82d199b50 sp=0xc82d199b30
net/http.serverHandler.ServeHTTP(0xc8202ae080, 0x7f7090b49ed8, 0xc828c93c70, 0xc820107b20)
        /usr/local/go/src/net/http/server.go:2081 +0x19e fp=0xc82d199bb0 sp=0xc82d199b50
net/http.(*conn).serve(0xc820bb0680)
        /usr/local/go/src/net/http/server.go:1472 +0xf2e fp=0xc82d199f78 sp=0xc82d199bb0
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc82d199f80 sp=0xc82d199f78
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2137 +0x44e
```

See https://weaveworks.slack.com/files/jcw/F0UGY0CUS/-.txt for full backtrace

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #1192

### CODE REVIEW

File: render/filters.go [hunk 1]
```diff
@@ -44,12 +44,13 @@ func ColorConnected(r Renderer) Renderer {
...
 
+			output := input.Copy()
 			for id := range connected {
-				input[id] = input[id].WithNode(report.MakeNodeWith(map[string]string{
+				output[id] = output[id].WithNode(report.MakeNodeWith(map[string]string{
 					IsConnected: "true",
 				}))
 			}
-			return input
+			return output
```
Comment (paulbellamy) [inline render/filters.go]: No need for the initial copy
```Go
output := RenderableNodes{}
for id := range connected {
  output[id] = input[id].WithNode(...) // I think this copies as well? check that.
}
return output
```

**Comment by paulbellamy** [comment]
Don't need to copy twice, but LGTM.

File: render/filters.go (hunk 1 above)
Comment (2opremio) [inline render/filters.go]: I am not sure it matters, but this is not equivalent, as output will only have the connected nodes.

Comment (paulbellamy) [inline render/filters.go]: ah, true, that does matter. But my point was more that doing `input.Copy()` first is copying all connected nodes twice.

Comment (2opremio) [inline render/filters.go]: Also, `output[id] = input[id].WithNode(...)` doesn't copy the values resulting in shared references (which may be fine, I am not sure)

Comment (2opremio) [inline render/filters.go]: It doesn't seem to be fine, I am getting errors in the tests.

**Pull request merged**